### PR TITLE
--optimized and --download flags

### PR DESCRIPTION
**Pull request by hemanth**
^

### CODE REVIEW

File: lib/output.js [hunk 1]
```diff
@@ -67,6 +71,14 @@ module.exports = function (parameters, response) {
     threshold
   ));
 
+  if (parameters.or) {
+    console.log('\nHere are your optimized images: ', optimizedResoruceURL);
+  }
+
+  if (parameters.download) {
+    got.stream(optimizedResoruceURL).pipe(createWriteStream('optimized.zip'));
+  }
```
Comment (hemanth) [inline lib/output.js]: `got` is writing a `zip` file which is `0kb` missing something? 
 //cc @sindresorhus

File: lib/output.js [hunk 2]
```diff
@@ -59,6 +62,7 @@ function getReporter(format) {
 module.exports = function (parameters, response) {
   var renderer = getReporter(parameters.format);
   var threshold = parameters.threshold || THRESHOLD;
+  var optimizedResoruceURL = RESOURCE_URL + 'url=' + response.id + '&strategy=' + parameters.strategy;
```
Comment (sindresorhus) [inline lib/output.js]: This should use `require('querystring').stringify`.

File: cli.js [hunk 3]
```diff
@@ -17,7 +17,9 @@ var cli = meow({
     '  --strategy     Strategy to use when analyzing the page: mobile|desktop',
     '  --format       Output format: cli|json|tap',
     '  --locale       Locale results should be generated in.',
-    '  --threshold    Threshold score to pass the PageSpeed test.'
+    '  --threshold    Threshold score to pass the PageSpeed test.',
+    '  --or           Get the URL of optimised resoruces.'
```
Comment (sindresorhus) [inline cli.js]: `--or` → `--optimized`

Comment (sindresorhus) [inline cli.js]: `optimised` → `optimized`

File: cli.js [hunk 4]
```diff
@@ -17,7 +17,9 @@ var cli = meow({
...
+    '  --download     Download optimised resources.'
```
Comment (sindresorhus) [inline cli.js]: `optimised` → `optimized`

File: index.js [hunk 5]
```diff
@@ -51,8 +53,6 @@ module.exports.output = function (url, opts, cb) {
       return;
     }
 
-    console.log(data);
```
Comment (sindresorhus) [inline index.js]: ?

File: index.js [hunk 6]
```diff
@@ -38,6 +39,7 @@ var psi = module.exports = function (url, opts, cb) {
 };
 
 module.exports.output = function (url, opts, cb) {
+
```
Comment (sindresorhus) [inline index.js]: no

File: lib/output.js [hunk 7]
```diff
@@ -2,7 +2,10 @@
 var prettyBytes = require('pretty-bytes');
 var sortOn = require('sort-on');
 var humanizeUrl = require('humanize-url');
+var got = require('got');
+var createWriteStream = require('fs').createWriteStream
```
Comment (sindresorhus) [inline lib/output.js]: semicolon

File: lib/output.js (hunk 1 above)
Comment (sindresorhus) [inline lib/output.js]: It's an async function and it's in a sync function. You'll need to convert the function to an async one with callback.

File: lib/output.js (hunk 2 above)
Comment (hemanth) [inline lib/output.js]: Yes, was planning for that.

File: cli.js (hunk 3 above)
Comment (hemanth) [inline cli.js]: Alright.

File: index.js (hunk 5 above)
Comment (hemanth) [inline index.js]: Why is that needed?

File: lib/output.js (hunk 1 above)
Comment (hemanth) [inline lib/output.js]: Yes, I do get the `async` part of it, but any other way here to pipe the stream, I could stick to `got.get(url, [options], [callback])`

File: index.js (hunk 5 above)
Comment (sindresorhus) [inline index.js]: Uhm, it logs the output. That's the whole point of that function.

File: lib/output.js (hunk 1 above)
Comment (sindresorhus) [inline lib/output.js]: I don't think you do. That wouldn't help. Again, nothing to do with `got`.

You need to convert [this function](https://github.com/addyosmani/psi/blob/96f9579ab5a0084ff20faab7048e1448a22154f5/lib/output.js#L59) into an async one and call the callback function when `got` is done. The reason it's 0kb is that the CLI calls `process.exit` before `got` has even gotten around to start downloading it.

Comment (hemanth) [inline lib/output.js]: Did get that `process.exit` is been invoked....hmm...I do get the async! Will add a callback to it.

Comment (hemanth) [inline lib/output.js]: Say:

```js
module.exports = function (parameters, response, cb) {
   //....
    if (parameters.download) {
      got.get(optimizedResoruceURL, function(err,data) {
         if(!err) return cb(null,data);
         return cb(err,null);
      });
  }
}
```
 And in `index.js`:

```js
output(handleOpts(url, opts), data, function(err, data) { if(!err) console.log(data);});
```

File: index.js (hunk 5 above)
Comment (hemanth) [inline index.js]: Oh, hmm, on CLI was seeing something like:

```sh
$ psi http://www.h3manth.com/  --strategy=desktop --download --optimized
{ kind: 'pagespeedonline#result',
  id: 'http://www.h3manth.com/',
  responseCode: 200,
  title: 'Hemanth.HM\'s paws on tech.',
  score: 92,
  pageStats: 
   { numberResources: 9,
     numberHosts: 4,
     totalRequestBytes: '890',
     numberStaticResources: 5,
     htmlResponseBytes: '4650',
     cssResponseBytes: '44909',
     imageResponseBytes: '104699',
     javascriptResponseBytes: '26623',
     otherResponseBytes: '41994',
     numberJsResources: 1,
     numberCssResources: 2 },
  formattedResults: 
   { locale: 'en_US',
     ruleResults: 
      { AvoidLandingPageRedirects: [Object],
        EnableGzipCompression: [Object],
        LeverageBrowserCaching: [Object],
        MainResourceServerResponseTime: [Object],
        MinifyCss: [Object],
        MinifyHTML: [Object],
        MinifyJavaScript: [Object],
        MinimizeRenderBlockingResources: [Object],
        OptimizeImages: [Object],
        PrioritizeVisibleContent: [Object] } },
  version: { major: 1, minor: 15 } }

--------------------------------------------------------
```

So muted it.

**Comment by hemanth** [comment]
@sindresorhus Done, we could use a spinner for download indication and also use a `--path` if the user wants to down the optimized data to particular path.

File: lib/output.js [hunk 8]
```diff
@@ -67,6 +72,20 @@ module.exports = function (parameters, response) {
     threshold
   ));
 
+  if (parameters.optimized) {
+    console.log('\nHere are your optimized images: ', humanizeUrl(optimizedResoruceURL));
+  }
+
+  if (parameters.download) {
+      new Download({mode: '755'})
```
Comment (sindresorhus) [inline lib/output.js]: Why did you switch out `got`? I see no reason to use `download` here as it has a very heavy dependency tree and `got` does exactly what we need.

File: index.js [hunk 9]
```diff
@@ -54,12 +54,12 @@ module.exports.output = function (url, opts, cb) {
     console.log(data);
 
     try {
-      output(handleOpts(url, opts), data);
+      output(handleOpts(url, opts), data, function() {
```
Comment (sindresorhus) [inline index.js]:
```js
output(handleOpts(url, opts), data, cb);
```

File: index.js [hunk 10]
```diff
@@ -54,12 +54,12 @@ module.exports.output = function (url, opts, cb) {
...
+        cb();
+      });
     } catch (err) {
       cb(err);
       return;
```
Comment (sindresorhus) [inline index.js]: this `return` can now be removed

File: lib/output.js [hunk 11]
```diff
@@ -56,9 +60,10 @@ function getReporter(format) {
   return require('./formats/' + format);
 }
 
-module.exports = function (parameters, response) {
+module.exports = function (parameters, response, cb) {
   var renderer = getReporter(parameters.format);
   var threshold = parameters.threshold || THRESHOLD;
+  var optimizedResoruceURL = RESOURCE_URL + querystring.stringify({url:response.id,'strategy': parameters.strategy})
```
Comment (sindresorhus) [inline lib/output.js]: no point in quoting `strategy`.

File: lib/output.js [hunk 12]
```diff
@@ -67,6 +72,20 @@ module.exports = function (parameters, response) {
...
+  if (parameters.download) {
+      new Download({mode: '755'})
+          .get(optimizedResoruceURL)
+          .dest('.')
+          .rename('./optimized.zip')
+          .run(function() {
+            cb();
+          });
+  }
+
```
Comment (sindresorhus) [inline lib/output.js]: The below error should now be passed in the callback:

File: index.js (hunk 10 above)
Comment (sindresorhus) [inline index.js]: actually, with https://github.com/addyosmani/psi/pull/59/files#r40241237 the try/catch can be removed

File: lib/output.js (hunk 12 above)
Comment (sindresorhus) [inline lib/output.js]: This also needs to be put before the download thingy.

File: lib/output.js (hunk 11 above)
Comment (hemanth) [inline lib/output.js]: I was expecting this comment ;)

File: lib/output.js (hunk 8 above)
Comment (hemanth) [inline lib/output.js]: Should I switch back to `got`?  I kina liked `Download` :)

File: index.js (hunk 9 above)
Comment (hemanth) [inline index.js]: :+1:

**Comment by hemanth** [comment]
@sindresorhus Done, thanks for the review.

File: lib/output.js (hunk 8 above)
Comment (sindresorhus) [inline lib/output.js]: Why do you need to rename the file? What's the original filename?

File: lib/output.js [hunk 13]
```diff
@@ -67,9 +72,23 @@ module.exports = function (parameters, response) {
...
 
+  if (parameters.optimized) {
+    console.log('\nHere are your optimized images: ', humanizeUrl(optimizedResoruceURL));
+  }
+
   if (response.score < threshold) {
     var err = new Error('Threshold of ' + threshold + ' not met with score of ' + response.score);
     err.noStack = true;
-    throw err;
+    throw new Error(err);
```
Comment (sindresorhus) [inline lib/output.js]: ?

Comment (sindresorhus) [inline lib/output.js]: From my earlier comment:

> The below error should now be passed in the callback instead of thrown.

File: lib/output.js [hunk 14]
```diff
@@ -67,9 +72,23 @@ module.exports = function (parameters, response) {
...
+  }
+
+  if (parameters.download) {
+      new Download({mode: '755'})
+          .get(optimizedResoruceURL)
+          .dest('.')
+          .rename('./optimized.zip')
+          .run(function() {
```
Comment (sindresorhus) [inline lib/output.js]: `.run(cb);`

File: lib/output.js (hunk 13 above)
Comment (hemanth) [inline lib/output.js]: There is a test case that is expecting a throw, shall I remove it?

**Comment by hemanth** [comment]
@sindresorhus Ping (:

File: lib/output.js (hunk 13 above)
Comment (sindresorhus) [inline lib/output.js]: No, change the test to fit the new async interface.

File: lib/output.js [hunk 15]
```diff
@@ -67,9 +72,21 @@ module.exports = function (parameters, response) {
...
+
   if (response.score < threshold) {
     var err = new Error('Threshold of ' + threshold + ' not met with score of ' + response.score);
     err.noStack = true;
-    throw err;
+    throw new Error(err);
+  }
+
+  if (parameters.download) {
+      new Download({mode: '755'})
```
Comment (sindresorhus) [inline lib/output.js]: Why `755`? I think the default is ok.

File: cli.js [hunk 16]
```diff
@@ -17,7 +17,9 @@ var cli = meow({
     '  --strategy     Strategy to use when analyzing the page: mobile|desktop',
     '  --format       Output format: cli|json|tap',
     '  --locale       Locale results should be generated in.',
-    '  --threshold    Threshold score to pass the PageSpeed test.'
+    '  --threshold    Threshold score to pass the PageSpeed test.',
+    '  --optimized    Get the URL of optimized resoruces.',
```
Comment (sindresorhus) [inline cli.js]: typo

File: README.md [hunk 17]
```diff
@@ -119,6 +119,8 @@ $ psi --help
     --format       Output format: cli|json|tap
     --locale       Locale results should be generated in.
     --threshold    Threshold score to pass the PageSpeed test.
+    --optimized    Get the URL of optimized resoruces.
```
Comment (sindresorhus) [inline README.md]: typo

**Comment by sindresorhus** [comment]
The flags also needs to be tested.

**Comment by sindresorhus** [comment]
@addyosmani You ok with these changes?

**Comment by addyosmani** [comment]
@sindresorhus I'm okay with these changes. @hemanth we've merged in the V2 updates. Could you rebase? We should be good to go once that's complete.

**Comment by hemanth** [comment]
@addyosmani Done.

**Comment by addyosmani** [comment]
LGTM. Thank you for the contributions :stars:

**Comment by hemanth** [comment]
Thank you. :+1:

**Comment by sindresorhus** [comment]
This was not ready to merge, though. Still missing tests as requested in https://github.com/addyosmani/psi/pull/59#issuecomment-143674942.

**Comment by addyosmani** [comment]
@sindresorhus Ugh. Review fail. It wasn't visible in the updated view here so I completely missed it. @hemanth We're going to hold off on cutting a new release on tests for this piece of work. Could you take a look at that in a follow-up PR next please?

**Comment by hemanth** [comment]
> The flags also needs to be tested.

Overlooked that comment. 

BTW as of now we have something like [test.js#L58](https://github.com/addyosmani/psi/blob/master/test/test.js#L58) options test case, test cases must be written for all of the flags?

@sindresorhus AKA review-bot! :wink:

**Comment by addyosmani** [comment]
Tests for just the two flags in this PR should be sufficient. We can review/release once those two land but currently holding off a release until those are in :star:

**Comment by hemanth** [comment]
We need to stub `console` here, are you ok to use `mocha/sinon` or something as simple as [test-console](https://github.com/jamesshore/test-console)?

**Comment by addyosmani** [comment]
Personally fine with mocha and mock-req unless @sindresorhus has a preference for something more lean.

**Comment by sindresorhus** [comment]
Sure

**Comment by hemanth** [comment]
I am down with fever :cactus: will be completing this in few days.

**Comment by hemanth** [comment]
@sindresorhus @addyosmani [Current status](https://github.com/hemanth/psi/blob/master/test/test.js#L76-L89) let me know if that looks good, will do a PR soon after.

Trying to figure out why `nock` says `[Error: Nock: No match for request GET ..`
